modules/common/src/main/scala/notebook/front/widgets/Magic.scala

```diff
@@ -89,6 +95,21 @@ object SamplerImplicits extends ExtraSamplerImplicits {
   implicit def MapSampler[K,V] = new Sampler[Map[K, V]] {
     def apply(x:Map[K,V], max:Int):Map[K,V] = x.toSeq.take(max).toMap
   }
+  /**
+   * count number of rows in a DataFrame, and do sampling if needed
+   */
+  implicit def AdvancedDataFrameSampler = new Sampler[DataFrame] {
+    def apply(df: DataFrame, max: Int): DataFrame = {
+      val count = df.cache().count()
+      if (count > max) {
+        println("DF is larger than maxRows. Will use sampling.")
```

Comment:
indeed, since the sample is passed at the render time, the message should be taken into account there, I'd say.

This gardian on the count you've made is a local check to avoid sampling of course, which is fair, and the log message is also fair however, I wouldn't leave the Sampler the charge to notify the view. my2¢ of course.